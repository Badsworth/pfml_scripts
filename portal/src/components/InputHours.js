import Dropdown from "./Dropdown";
import Fieldset from "./Fieldset";
import FormLabel from "./FormLabel";
import InputText from "./InputText";
import PropTypes from "prop-types";
import React from "react";
import classnames from "classnames";
import convertMinutesToHours from "../utils/convertMinutesToHours";
import { range } from "lodash";

const InputHours = (props) => {
  const formGroupClasses = classnames("usa-form-group", {
    "usa-form-group--error": !!props.errorMsg,
  });
  const inputFormGroupClasses = {
    hours: classnames("display-inline-block margin-right-3", {
      "margin-top-neg-105": props.smallLabel,
    }),
    minutes: classnames("display-inline-block", {
      "margin-top-neg-105": props.smallLabel,
    }),
  };

  const inputClasses = {
    hours: classnames({ "usa-input--error": props.hoursInvalid }),
    minutes: classnames("width-15", {
      "usa-input--error": props.minutesInvalid,
    }),
  };

  const minuteChoices = range(60 / props.minutesIncrement).map((i) => ({
    label: String(i * props.minutesIncrement),
    value: i * props.minutesIncrement,
  }));

  const value = convertMinutesToHours(props.value || 0);

  if (value.minutes % props.minutesIncrement !== 0) {
    // eslint-disable-next-line no-console
    console.warn(
      `Minutes value of ${props.value} is not a multipe of provided minutesIncrement (${props.minutesIncrement}).`
    );
  }

  const handleHoursChange = (event) =>
    dispatchChange(Math.floor(event.target.value) * 60 + value.minutes, event);

  const handleMinutesChange = (event) =>
    dispatchChange(value.hours * 60 + Number(event.target.value), event);

  /**
   * Call props.onChange with an argument value in a shape resembling Event so
   * that our form event handlers can manage this field's state just like
   * it does with other fields like InputText. We also include the original
   * event, but only for debugging purposes.
   * @param {string} value - ISO 8601 date string
   * @param {SyntheticEvent} originalEvent - Original event that triggered this change
   */
  function dispatchChange(value, originalEvent) {
    props.onChange({
      _originalEvent: originalEvent,
      target: {
        name: props.name,
        type: "numeric",
        value,
      },
    });
  }

  return (
    <Fieldset className={formGroupClasses}>
      <FormLabel
        component="legend"
        errorMsg={props.errorMsg}
        small={props.smallLabel}
        example={props.example}
        hint={props.hint}
        weight={props.labelWeight}
        optionalText={props.optionalText}
      >
        {props.label}
      </FormLabel>
      <InputText
        formGroupClassName={inputFormGroupClasses.hours}
        inputClassName={inputClasses.hours}
        name="hours"
        label={props.hoursLabel}
        smallLabel
        width="small"
        labelWeight="normal"
        inputMode="numeric"
        pattern="[0-9]*"
        value={value.hours}
        onChange={handleHoursChange}
      />
      <Dropdown
        formGroupClassName={inputFormGroupClasses.minutes}
        selectClassName={inputClasses.minutes}
        name="minutes"
        label={props.minutesLabel}
        smallLabel
        labelWeight="normal"
        choices={minuteChoices}
        value={value.minutes}
        onChange={handleMinutesChange}
      />
    </Fieldset>
  );
};

InputHours.propTypes = {
  /**
   * HTML input `name` attribute
   */
  name: PropTypes.string.isRequired,
  /**
   * Localized error message. Setting this enables the error state styling.
   */
  errorMsg: PropTypes.node,
  /**
   * Localized label for the entire fieldset
   */
  label: PropTypes.node.isRequired,
  /**
   * Override the default label font weight
   */
  labelWeight: PropTypes.oneOf(["bold", "normal"]),
  /**
   * Enable the smaller label variant
   */
  smallLabel: PropTypes.bool,
  /**
   * Localized example text
   */
  example: PropTypes.string,
  /**
   * Localized hint text
   */
  hint: PropTypes.node,
  /**
   * Localized text indicating this field is optional
   */
  optionalText: PropTypes.node,
  /**
   * Localized label for the hours field
   */
  hoursLabel: PropTypes.string,
  /**
   * Apply error styling to the day `input`
   */
  hoursInvalid: PropTypes.bool,
  /**
   * Localized label for the minutes field
   */
  minutesLabel: PropTypes.string,
  /**
   * Apply error styling to the day `input`
   */
  minutesInvalid: PropTypes.bool,
  /**
   * Amount to increment minutes input. Must be an integer and must be a factor of 60.
   */
  minutesIncrement: PropTypes.number.isRequired,
  /**
   * HTML input `onChange` attribute
   */
  onChange: PropTypes.func.isRequired,
  /**
   * Hours value represented in minutes. Must be a whole number.
   */
  value: PropTypes.number,
};

export default InputHours;
