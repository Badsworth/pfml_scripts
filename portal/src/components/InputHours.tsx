import { isNil, range } from "lodash";
import Dropdown from "./Dropdown";
import Fieldset from "./Fieldset";
import FormLabel from "./FormLabel";
import InputNumber from "./InputNumber";
import React from "react";
import classnames from "classnames";
import convertMinutesToHours from "../utils/convertMinutesToHours";

interface InputHoursProps {
  /**
   * HTML input `name` attribute
   */
  name: string;
  /**
   * Localized error message. Setting this enables the error state styling.
   */
  errorMsg?: React.ReactNode;
  /**
   * Localized label for the entire fieldset
   */
  label: React.ReactNode;
  /**
   * Override the label's default text-bold class
   */
  labelClassName?: string;
  /**
   * Enable the smaller label variant
   */
  smallLabel?: boolean;
  /**
   * Localized example text
   */
  example?: string;
  /**
   * Localized hint text
   */
  hint?: React.ReactNode;
  /**
   * Localized text indicating this field is optional
   */
  optionalText?: React.ReactNode;
  /**
   * Localized label for the hours field
   */
  hoursLabel?: string;
  /**
   * Apply error styling to the day `input`
   */
  hoursInvalid?: boolean;
  /**
   * Localized label for the minutes field
   */
  minutesLabel?: string;
  /**
   * Apply error styling to the day `input`
   */
  minutesInvalid?: boolean;
  /**
   * Amount to increment minutes input. Must be an integer and must be a factor of 60.
   */
  minutesIncrement: number;
  /**
   * HTML input `onChange` attribute
   */
  onChange: React.ChangeEventHandler<HTMLInputElement>;
  /**
   * Hours value represented in minutes. Must be a whole number.
   */
  value?: number | string;
}

const InputHours = (props: InputHoursProps) => {
  const formGroupClasses = classnames("usa-form-group", {
    "usa-form-group--error": !!props.errorMsg,
  });
  const inputFormGroupClasses = {
    hours: classnames("display-inline-block margin-right-3", {
      "margin-top-105": props.smallLabel,
    }),
    minutes: classnames("display-inline-block", {
      "margin-top-105": props.smallLabel,
    }),
  };

  const inputClasses = {
    hours: classnames({ "usa-input--error": props.hoursInvalid }),
    minutes: classnames("width-15", {
      "usa-input--error": props.minutesInvalid,
    }),
  };

  const minuteChoices = range(60 / props.minutesIncrement).map((i) => ({
    label: String(i * props.minutesIncrement),
    value: i * props.minutesIncrement,
  }));

  const hoursMinutes = convertMinutesToHours(Number(props.value) || 0);

  if (hoursMinutes.minutes % props.minutesIncrement !== 0) {
    // eslint-disable-next-line no-console
    console.warn(
      `Minutes value of ${props.value} is not a multiple of provided minutesIncrement (${props.minutesIncrement}).`
    );
  }

  const handleHoursChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    // Change input to null if the hours field is blank and no minutes are provided
    const isBlank = event.target.value === "" && hoursMinutes.minutes === 0;
    const value = isBlank
      ? null
      : Math.floor(Number(event.target.value)) * 60 + hoursMinutes.minutes;
    dispatchChange(value);
  };

  const handleMinutesChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    // Change input to null if the minutes field is blank and no hours are provided
    const isBlank = event.target.value === "" && hoursMinutes.hours === 0;
    const value = isBlank
      ? null
      : hoursMinutes.hours * 60 + Number(event.target.value);
    dispatchChange(value);
  };

  /**
   * Call props.onChange with an argument value in a shape resembling Event so
   * that our form event handlers can manage this field's state just like
   * it does with other fields like InputText. We also include the original
   * event, but only for debugging purposes.
   */
  function dispatchChange(value: null | number) {
    const target = document.createElement("input");
    const parsedValue = value === null ? "" : value.toString();
    target.setAttribute("name", props.name);
    target.setAttribute("value", parsedValue);
    target.setAttribute("data-value-type", "integer");
    target.name = props.name;
    target.value = parsedValue;

    props.onChange({
      target,
    } as React.ChangeEvent<HTMLInputElement>);
  }

  return (
    <Fieldset className={formGroupClasses}>
      <FormLabel
        component="legend"
        errorMsg={props.errorMsg}
        small={props.smallLabel}
        example={props.example}
        hint={props.hint}
        labelClassName={props.labelClassName}
        optionalText={props.optionalText}
      >
        {props.label}
      </FormLabel>
      <InputNumber
        formGroupClassName={inputFormGroupClasses.hours}
        inputClassName={inputClasses.hours}
        name="hours"
        label={props.hoursLabel}
        smallLabel
        width="small"
        labelClassName="text-normal"
        inputMode="numeric"
        value={isNil(props.value) ? "" : hoursMinutes.hours}
        valueType="integer"
        allowNegative={false}
        onChange={handleHoursChange}
      />
      <Dropdown
        formGroupClassName={inputFormGroupClasses.minutes}
        selectClassName={inputClasses.minutes}
        name="minutes"
        label={props.minutesLabel}
        smallLabel
        labelClassName="text-normal"
        choices={minuteChoices}
        value={isNil(props.value) ? "" : hoursMinutes.minutes}
        onChange={handleMinutesChange}
      />
    </Fieldset>
  );
};

export default InputHours;
