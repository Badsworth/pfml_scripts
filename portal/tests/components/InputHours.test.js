import { mount, shallow } from "enzyme";
import InputHours from "../../src/components/InputHours";
import React from "react";

function render(customProps = {}, mountComponent = false) {
  const props = Object.assign(
    {
      label: "Label",
      hoursLabel: "Hours Label",
      minutesLabel: "Minutes Label",
      name: "field-name",
      minutesIncrement: 15,
      emptyMinutesLabel: "Select minute",
      onChange: jest.fn(),
    },
    customProps
  );

  const component = <InputHours {...props} />;

  return {
    props,
    wrapper: mountComponent ? mount(component) : shallow(component),
  };
}

describe("InputHours", () => {
  it("renders component", () => {
    const { wrapper } = render({ value: 40 * 60 + 15 });

    expect(wrapper).toMatchSnapshot();
  });

  it("tightens label spacing if smallLabel prop is true", () => {
    const { wrapper } = render({ smallLabel: true });

    expect(
      wrapper
        .find("InputText")
        .props()
        .formGroupClassName.includes("margin-top-neg-105")
    ).toBe(true);
    expect(
      wrapper
        .find("Dropdown")
        .props()
        .formGroupClassName.includes("margin-top-neg-105")
    ).toBe(true);
  });

  it("passes errorMsg to FormLabel if errorMsg is set", () => {
    const { wrapper } = render({ errorMsg: "Something went wrong" });

    expect(wrapper.find("FormLabel").props().errorMsg).toEqual(
      "Something went wrong"
    );
    expect(wrapper.find("Fieldset").hasClass("usa-form-group--error")).toBe(
      true
    );
  });

  it("adds error class to the day input component when hoursInvalid is true", () => {
    const { wrapper } = render({ hoursInvalid: true });

    expect(
      wrapper
        .find("InputText")
        .props()
        .inputClassName.includes("usa-input--error")
    ).toBe(true);
  });

  it("adds error class to the month input component when minutesInvalid is true", () => {
    const { wrapper } = render({ minutesInvalid: true });

    expect(
      wrapper
        .find("Dropdown")
        .props()
        .selectClassName.includes("usa-input--error")
    ).toBe(true);
  });

  it("renders correctly if there are 0 hours", () => {
    const { wrapper } = render({ value: 15 });

    expect(wrapper.find("InputText").props().value).toEqual(0);
    expect(wrapper.find("Dropdown").props().value).toEqual(15);
  });

  it("renders correctly with 0 minutes", () => {
    const { wrapper } = render({ value: 60 });
    expect(wrapper.find("InputText").props().value).toEqual(1);
    expect(wrapper.find("Dropdown").props().value).toEqual(0);
  });

  it("renders correctly if both hours and minutes are 0", () => {
    const { wrapper } = render({ value: 0 });
    expect(wrapper.find("InputText").props().value).toEqual(0);
    expect(wrapper.find("Dropdown").props().value).toEqual(0);
  });

  it("renders empty hours and minutes if value is undefined", () => {
    const { wrapper } = render({ value: undefined });
    expect(wrapper.find("InputText").props().value).toEqual("");
    expect(wrapper.find("Dropdown").props().value).toEqual("");
  });

  it("logs warning if minutesIncrements is not a multiple of value", () => {
    const warnSpy = jest
      .spyOn(console, "warn")
      .mockImplementationOnce(() => {});

    const { wrapper } = render({ value: 33, minutesIncrement: 15 });
    expect(wrapper.find("InputText").props().value).toEqual(0);
    expect(wrapper.find("Dropdown").props().value).toEqual(33);
    expect(warnSpy).toHaveBeenCalled();
  });

  it("renders empty hours if hours value is erased and minutes value is 0", () => {
    const { wrapper, props } = render({ value: 8 * 60 });

    wrapper.find("InputText").simulate("change", {
      target: { value: "" },
    });

    expect(props.onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: { name: props.name, type: "numeric", value: null },
      })
    );
  });

  it("sets hours to 0 if hours are erased and minutes are greater than 0", () => {
    const { wrapper, props } = render({ value: 8 * 60 + 15 });

    wrapper.find("InputText").simulate("change", {
      target: { value: "" },
    });

    expect(props.onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: { name: props.name, type: "numeric", value: 0 * 60 + 15 },
      })
    );
  });

  it("renders empty hours if minutes value is erased and hours value is 0", () => {
    const { wrapper, props } = render({ value: 15 });

    wrapper.find("Dropdown").simulate("change", {
      target: { value: "" },
    });

    expect(props.onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: { name: props.name, type: "numeric", value: null },
      })
    );
  });

  it("sets minutes to 0 if minutes are erased and hours are greater than 0", () => {
    const { wrapper, props } = render({ value: 8 * 60 + 15 });

    wrapper.find("Dropdown").simulate("change", {
      target: { value: "" },
    });

    expect(props.onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: { name: props.name, type: "numeric", value: 8 * 60 + 0 },
      })
    );
  });
});
