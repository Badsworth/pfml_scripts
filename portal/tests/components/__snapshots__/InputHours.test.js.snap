// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputHours renders component 1`] = `
<Fieldset
  className="usa-form-group"
>
  <FormLabel
    component="legend"
  >
    Label
  </FormLabel>
  <InputText
    formGroupClassName="display-inline-block margin-right-3"
    inputClassName=""
    inputMode="numeric"
    label="Hours Label"
    labelClassName="text-normal"
    name="hours"
    onChange={[Function]}
    smallLabel={true}
    value={40}
    valueType="integer"
    width="small"
  />
  <Dropdown
    choices={
      Array [
        Object {
          "label": "0",
          "value": 0,
        },
        Object {
          "label": "15",
          "value": 15,
        },
        Object {
          "label": "30",
          "value": 30,
        },
        Object {
          "label": "45",
          "value": 45,
        },
      ]
    }
    formGroupClassName="display-inline-block"
    label="Minutes Label"
    labelClassName="text-normal"
    name="minutes"
    onChange={[Function]}
    selectClassName="width-15"
    smallLabel={true}
    value={15}
  />
</Fieldset>
`;
